### PR TITLE
Added /etc/ci/chaos_monkey_test.py.

### LINKED ISSUE #10568: Add wpt tests which test pipeline closure
We now have `--random-pipeline-closure-probability`, so we should have tests which use it!

### PR DESCRIPTION
**Pull request by asajeffrey**
Add a test to `/etc/ci` which runs a subset of `test-wpt` with `--random-pipeline-failure-probability=0.2`, and checks to make sure that there's no `CRASH` reports, so the constellation survived the experience, even if a lot of tests failed.

IRC conversation at http://logs.glob.uno/?c=mozilla%23servo&s=27+Apr+2016&e=27+Apr+2016#c416510

Fixes #10568.

r? @aneeshusa

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10899)

### CODE REVIEW

**Comment by asajeffrey** [comment]
Fixes #10568

File: etc/ci/chaos_monkey_test.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+# by random pipeline closure, so we ignore the status code
+# returned by the test command (which is why we can't use check_output).
+
+TEST_RESULTS = Popen(TEST_CMD, stdout=PIPE)
+TEST_CRASHES = False
+TEST_STDOUT = {}
+
+for line in TEST_RESULTS.stdout:
+    report = json.loads(line.decode('utf-8'))
+    if report.get("action", None) == "process_output":
```
Comment (frewsxcv) [inline etc/ci/chaos_monkey_test.py]: The `default` value for `dict.get` is `None`, so explicitly specifying isn't necessary

File: etc/ci/chaos_monkey_test.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        print(report.get("thread", None) + " - " + report.get("data", None))
+    status = report.get("status", None)
+    if status:
+        print(report.get("thread", None) + " - " + status + " - " + report.get("test", None))
+        TEST_CRASHES = TEST_CRASHES or (status == "CRASH")
+
+if TEST_CRASHES:
+    exit(1)
```
Comment (frewsxcv) [inline etc/ci/chaos_monkey_test.py]: You probably want [`sys.exit`](http://stackoverflow.com/questions/6501121/the-difference-between-exit-and-sys-exit-in-python)

**Comment by asajeffrey** [comment]
Fixed and squashed. You can tell my python is a bit rusty.

File: etc/ci/chaos_monkey_test.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+# <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+
+import json
+import sys
+from subprocess import Popen, PIPE
+
+
+def is_crash(report):
```
Comment (aneeshusa) [inline etc/ci/chaos_monkey_test.py]: This method isn't being used, so it can be removed.

File: etc/ci/chaos_monkey_test.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    "content-security-policy"
+]
+
+# Note that there will probably be test failures caused
+# by random pipeline closure, so we ignore the status code
+# returned by the test command (which is why we can't use check_output).
+
+TEST_RESULTS = Popen(TEST_CMD, stdout=PIPE)
+TEST_CRASHES = False
+TEST_STDOUT = {}
```
Comment (aneeshusa) [inline etc/ci/chaos_monkey_test.py]: None of these three are constants, so they should be in lowercase `camel_case`.

File: etc/ci/chaos_monkey_test.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# Copyright 2013 The Servo Project Developers. See the COPYRIGHT
+# file at the top-level directory of this distribution.
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+# <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+
+import json
```
Comment (aneeshusa) [inline etc/ci/chaos_monkey_test.py]: Let's add `from __future__ import absolute_import, print_function` above this (with an empty line between that an import json) .

File: etc/ci/chaos_monkey_test.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+# returned by the test command (which is why we can't use check_output).
+
+TEST_RESULTS = Popen(TEST_CMD, stdout=PIPE)
+TEST_CRASHES = False
+TEST_STDOUT = {}
+
+for line in TEST_RESULTS.stdout:
+    report = json.loads(line.decode('utf-8'))
+    if report.get("action") == "process_output":
+        print(report.get("thread") + " - " + report.get("data"))
```
Comment (aneeshusa) [inline etc/ci/chaos_monkey_test.py]: Let's use `str.format` to make these nicer: 
```
print("{} - {}".format(report.get('thread'), report.get('data')))
```

File: etc/ci/chaos_monkey_test.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    status = report.get("status")
+    if status:
+        print(report.get("thread") + " - " + status + " - " + report.get("test"))
```
Comment (aneeshusa) [inline etc/ci/chaos_monkey_test.py]: Ditto about using `str.format`.

**Comment by aneeshusa** [comment]
Which outputs of this do you need? We can gate on the exit code, and Buildbot will automatically pick up on anything we `print()` to stdout. If you want to capture the `chaos-monkey.log`, we'll need to make that explicit when adding it to steps.yml (or send it to stderr, which should get picked up automatically).

I looked at the output and there's quite a bit of it - do we only want to log the CRASHes to output, or is having the rest of it useful as well?

File: etc/ci/chaos_monkey_test.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        TEST_CRASHES = TEST_CRASHES or (status == "CRASH")
```
Comment (aneeshusa) [inline etc/ci/chaos_monkey_test.py]: style nit: I find this easier to read, but up to you:
```
if status == "CRASH":
    TEST_CRASHES = True
```

**Comment by aneeshusa** [comment]
Will we ever have both `action == 'process_output'` and a `status`? Not sure if you want to avoid printing two lines in that case.

File: etc/ci/chaos_monkey_test.py (hunk 3 above)
Comment (asajeffrey) [inline etc/ci/chaos_monkey_test.py]: Fixed.

File: etc/ci/chaos_monkey_test.py (hunk 4 above)
Comment (asajeffrey) [inline etc/ci/chaos_monkey_test.py]: Fixed.

File: etc/ci/chaos_monkey_test.py (hunk 5 above)
Comment (asajeffrey) [inline etc/ci/chaos_monkey_test.py]: Done.

File: etc/ci/chaos_monkey_test.py (hunk 6 above)
Comment (asajeffrey) [inline etc/ci/chaos_monkey_test.py]: Done.

File: etc/ci/chaos_monkey_test.py (hunk 7 above)
Comment (asajeffrey) [inline etc/ci/chaos_monkey_test.py]: Done.

**Comment by asajeffrey** [comment]
IRC chat about outputs: http://logs.glob.uno/?c=mozilla%23servo&s=28+Apr+2016&e=28+Apr+2016#c417879

TL;DR: don't capture the whole log.

**Comment by asajeffrey** [comment]
`action` and `status` are exclusive, even if we do get a line with both, it's worth printing out both lots of info.

**Comment by asajeffrey** [comment]
Fixed and squashed.

**Comment by asajeffrey** [comment]
I also increased the probability of pipeline closure to 0.1, to make sure that some crashes happen!

File: etc/ci/chaos_monkey_test.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    # cross-origin iframes, which are good for stress-testing pipelines
+    "content-security-policy"
+]
+
+# Note that there will probably be test failures caused
+# by random pipeline closure, so we ignore the status code
+# returned by the test command (which is why we can't use check_output).
+
+test_results = Popen(TEST_CMD, stdout=PIPE)
+test_crashes = False
```
Comment (aneeshusa) [inline etc/ci/chaos_monkey_test.py]: Quick nit: Let's change this to `any_crashes` because `if any_crashes` reads better.

Comment (asajeffrey) [inline etc/ci/chaos_monkey_test.py]: Sure. I'll do this tomorrow, when I'm at a computer.

**Comment by asajeffrey** [comment]
Fixed and squashed.

**Comment by aneeshusa** [comment]
@bors-servo r+

We should file a follow-up issue for gating on this in the CI.

**Comment by asajeffrey** [comment]
Filed #10927
